Add RP2 DVI HSTX driver

sfe-SparkFro opened the pull request.

sfe-SparkFro committed:
Add initial RP2350 HSTX DVI driver

sfe-SparkFro committed:
Clean up RP2 DVI HSTX driver.

Inherit from CV2_Display
Properly implement imshow() and clear()
Add gray8 and bgra8888 conversion functions to cv2_display.py
Import from displays/__init__.py
Move test code to new example

gigapod reviewed (approved):
No obvious issues that I see.

Well structure/commented code. Bonus points for ASCII diagrams!

malcolm-sparkfun commented on red_vision/displays/dvi_rp2_hstx.py:
```diff
@@ -0,0 +1,1110 @@
+        # +---------------------------+
+        #            RP2350
+        # 
+        # The control blocks can change the executer's read and wite addresses,
```
typo: wite for write

sfe-SparkFro replied:
Thanks, fixed!

malcolm-sparkfun commented on red_vision/displays/dvi_rp2_hstx.py:
```diff
@@ -0,0 +1,1110 @@
+            # registers; the top 16 bits must be `0xACCE`, otherwise a bus
+            # fault occurs, causing the processor to freeze... So we need to
+            # add those bits.
+            ACCESSCTRL_PASSWORD_BITS = 0xACCE0000
```
Good find on this password thing. I feel like that would have taken me forever to find out haha...

malcolm-sparkfun commented on red_vision/displays/dvi_rp2_hstx.py:
```diff
@@ -0,0 +1,1110 @@
+        #              |                     +->| Dispatcher | |
+        #              |                        +------------+ |
+        #              +---------------------------------------+
+        #                               RP2350
```
Good description and picture.

malcolm-sparkfun commented on red_vision/displays/dvi_rp2_hstx.py:
```diff
@@ -0,0 +1,1110 @@
+            # executer will write the nested `_cb_streamer_abort_nested` control
+            # block to the CHAN_ABORT register, aborting the streamer DMA.
+            DMA_BASE = 0x50000000
+            DMA_CHAN_ABORT_ADDR = DMA_BASE + 0x464
```
Nitpick, but since you do this all in one line below (i.e. `XIP_STREAM_ADDR = 0x400C8000 + 0x14`) and since you don't use this `DMA_BASE` anywhere else, you could consider doing this all in one line instead of making the separate `DMA_BASE` variable.

sfe-SparkFro replied:
Decided to make all of them have the format of:

```
xyz_BASE = 0x12345678
register = xyz_BASE + 0x12
```

And all matching the register names in the RP2350 datasheet as best as possible. Although the `xyz_BASE` values are only used once, IMO it's a bit more readable and followable if you've got the datasheet pulled up.

malcolm-sparkfun replied:
Agreed

malcolm-sparkfun reviewed (approved):
I left a couple of minor nits that you can choose to disregard at your leisure.

This is a very impressive piece of code :)

sfe-SparkFro committed:
Set HSTX pin mode with `Pin.ALT_HSTX`

In anticipation of https://github.com/micropython/micropython/pull/18458,
which is slated for the v1.27.0 release.

sfe-SparkFro committed:
Fix typo

sfe-SparkFro committed:
Update register names to better match datasheet
